fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/BoundRecordQuery.java

```diff
@@ -0,0 +1,172 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Streams;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hasher;
+import com.google.common.hash.Hashing;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.Nonnull;
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static com.apple.foundationdb.record.query.expressions.BooleanComponent.groupedComparisons;
+import static com.apple.foundationdb.record.query.expressions.BooleanComponent.pairwiseDependentComparisons;
+
+/**
+ * A bound record query wraps an instance of {@link RecordQuery} and {@link RecordStoreState} in order to associate
+ * the query with the state of a store. That allows for differentiation of queries depending on the store state, e.g.
+ * the availability of indexes at the time the query was bound to the state, etc.
+ *
+ * Bound record queries also define {@link #hashCode()} and {@link #equals(Object)} in a way that allows for queries to
+ * be considered compatible to each other which is useful with respect to pre-bound parameter markers.
+ *
+ * Two queries are considered compatible, iff
+ * 1. their query structure is equal (minus their pre-bound parameter markers)
+ * 2. their associated record store state was compatible at the time of binding the {@link RecordQuery} to the
+ *    {@link RecordStoreState}
+ * 3. their pre-bound parameter markers use compatible pre-bound values
+ */
+public class BoundRecordQuery {
+    @Nonnull
+    private final RecordStoreState recordStoreState;
+    @Nonnull
+    private final RecordQuery recordQuery;
+    @Nonnull
+    private final Supplier<Set<String>> parametersSupplier;
+    @Nonnull
+    private final Supplier<Integer> hashCodeSupplier;
+
+    public BoundRecordQuery(@Nonnull final RecordStoreState recordStoreState, @Nonnull final RecordQuery recordQuery) {
+        this.recordStoreState = recordStoreState;
+        this.recordQuery = recordQuery;
+        this.parametersSupplier = Suppliers.memoize(this::computeParameters);
+        this.hashCodeSupplier = Suppliers.memoize(this::computeHashCode);
+    }
+
+    @Nonnull
+    public RecordStoreState getRecordStoreState() {
+        return recordStoreState;
+    }
+
+    @Nonnull
+    public RecordQuery getRecordQuery() {
+        return recordQuery;
+    }
+
+    public Set<String> getParameters() {
+        return parametersSupplier.get();
+    }
+
+    @Override
+    public int hashCode() {
+        return hashCodeSupplier.get();
+    }
+
+    @SuppressWarnings("UnstableApiUsage")
+    private int computeHashCode() {
+        final HashFunction hashFunction = Hashing.murmur3_32();
+        final Hasher hasher = hashFunction.newHasher();
+        hasher.putInt(recordQuery.hashCode());
+        hasher.putInt(recordStoreState.getIndexStates().hashCode());
```

Comment:
I think this may also want to include the metadata version from the store header. I'm pretty sure the index states will only contain the state of non-readable indexes, all readable indexes are implicitly readable by the nature of the fact that they are not in this list.  Thus, you could have the scenario in which the cached plan had, say, all the indexes readable, in the meantime a new index was added and built for a given record store, so the state of the store is that all the indexes are readable so it computes the same hash.

Reply:
I think we decided that something at a higher level would take account of the meta-data version, since it also needs to account for entirely different meta-data trains ("catalog").